cardano-crypto-class/src/Cardano/Crypto/EllipticCurve/BLS12_381/Internal.hs

```diff
@@ -0,0 +1,875 @@
+
+sizeScalar :: forall i. Num i => i
+sizeScalar = fromIntegral c_size_blst_scalar
+
+newtype Scalar = Scalar (ForeignPtr Void)
+
+withIntScalar :: Integer -> (ScalarPtr -> IO a) -> IO a
+withIntScalar i go = do
+  s <- scalarFromInteger i
+  withScalar s go
+
+withScalar :: Scalar -> (ScalarPtr -> IO a) -> IO a
+withScalar (Scalar p2) go = do
+  withForeignPtr p2 (go . ScalarPtr)
+
+withNewScalar :: (ScalarPtr -> IO a) -> IO (a, Scalar)
+withNewScalar go = do
+  p2 <- mallocForeignPtrBytes sizeScalar
+  x <- withForeignPtr p2 (go . ScalarPtr)
+  return (x, Scalar p2)
+
+withNewScalar_ :: (ScalarPtr -> IO a) -> IO a
+withNewScalar_ = fmap fst . withNewScalar
+
+withNewScalar' :: (ScalarPtr -> IO a) -> IO Scalar
+withNewScalar' = fmap snd . withNewScalar
+
+cloneScalar :: Scalar -> IO Scalar
+cloneScalar (Scalar a) = do
+  b <- mallocForeignPtrBytes sizeScalar
+  withForeignPtr a $ \ap ->
+    withForeignPtr b $ \bp ->
+      copyBytes bp ap sizeScalar
+  return (Scalar b)
+
+sizeFr :: forall i. Num i => i
+sizeFr = fromIntegral c_size_blst_fr
+
+newtype Fr = Fr (ForeignPtr Void)
+
+withFr :: Fr -> (FrPtr -> IO a) -> IO a
+withFr (Fr p2) go = do
+  withForeignPtr p2 (go . FrPtr)
+
+withNewFr :: (FrPtr -> IO a) -> IO (a, Fr)
+withNewFr go = do
+  p2 <- mallocForeignPtrBytes sizeFr
+  x <- withForeignPtr p2 (go . FrPtr)
+  return (x, Fr p2)
+
+withNewFr_ :: (FrPtr -> IO a) -> IO a
+withNewFr_ = fmap fst . withNewFr
+
+withNewFr' :: (FrPtr -> IO a) -> IO Fr
+withNewFr' = fmap snd . withNewFr
+
+cloneFr :: Fr -> IO Fr
+cloneFr (Fr a) = do
+  b <- mallocForeignPtrBytes sizeFr
+  withForeignPtr a $ \ap ->
+    withForeignPtr b $ \bp ->
+      copyBytes bp ap sizeFr
+  return (Fr b)
+
+scalarToInteger :: Scalar -> IO Integer
+scalarToInteger scalar = withScalar scalar $ \scalarPtr -> do
+  allocaBytes sizeScalar $ \rawPtr -> do
+    c_blst_bendian_from_scalar rawPtr scalarPtr
+    cstrToInteger rawPtr sizeScalar
+
+cstrToInteger :: Ptr CChar -> Int -> IO Integer
+cstrToInteger p l = do
+  go l (castPtr p)
+  where
+    go :: Int -> Ptr CUChar -> IO Integer
+    go 0 _ = return 0
+    go n ptr = do
+      val <- peek ptr
+      res <- go (pred n) (plusPtr ptr 1)
+      return $ res .|. shiftL (fromIntegral val) (8 * pred n)
+
+integerToBS :: Integer -> ByteString
+integerToBS 0 = BS.empty
+integerToBS n
+  | n < 0
+  = error "Cannot convert negative Integer to ByteString"
+  | otherwise
+  = BS.snoc
+      (integerToBS (n `shiftR` 8))
+      (fromIntegral n)
+
+padBS :: Int -> ByteString -> ByteString
+padBS i b
+  | i > BS.length b
+  = BS.replicate (i - BS.length b) 0 <> b
+  | otherwise
+  = b
+
+integerAsCStrL :: Int -> Integer -> (Ptr CChar -> Int -> IO a) -> IO a
+integerAsCStrL i n f = do
+  let bs = padBS i $ integerToBS n
+  BS.useAsCStringLen bs $ uncurry f
+
+scalarFromInteger :: Integer -> IO Scalar
+scalarFromInteger n = do
+  withNewScalar' $ \scalarPtr -> do
+    integerAsCStrL sizeScalar (n `mod` scalarPeriod) $ \str _length -> do
```

Comment:
We either need to check that `n` is not negative or use `rem` here instead:
```suggestion
    integerAsCStrL sizeScalar (n `rem` scalarPeriod) $ \str _length -> do
```

This has to do with the way modulo arithmetics work on the negative numbers. Here is an example:
```haskell
> scalarPeriod = 0x73eda753299d7d483339d80809a1d80553bda402fffe5bfeffffffff00000001 :: Integer
> (1 - scalarPeriod) `mod` scalarPeriod
1
> (1 - scalarPeriod) `rem` scalarPeriod
-52435875175126190479447740508185965837690552500527637822603658699938581184512
```

Reply:
what's wrong with 1 - scalarPeriod `mod` scalarPeriod being equal to 1? That's what we'd actually expect, right?

Reply:
So, you are saying negative values are allowed as input? If that's the case then it's fine, I just would not have expect it to be the case. This sort of overflows is the constant source of bugs in the bounded integer arithmetic, that's why I always try to be extra careful with this. I am not familiar with BLS12 at all, so it might be totally expected by the users of the library.

Maybe it would be good a idea to add a note about negative numbers on the `mult` function? It is sort of implied by the haddock: "The scalar will be brought into the range of modular arithmetic by means of a modulo operation over the 'scalarPeriod'", but being more explicit about negative numbers could be useful IMHO.

Reply:
Sometimes, when interacting with elliptic curve points we want to check things such as `a * A - b * B` (where capital letters represent points). A way to do this would be `a * A + NEG(b * B)`, but if we allow negative numbers, we allow the user to call `a * A + (-b) * B`. 

However, Now that I'm thinking about it, the `NEG` operation on an ecpoint might be cheaper (it's a sign bit flip only). I'll think about this. We have to change the representation of points nonetheless, because @kwxm found a bug that the following:  `neg(A) =? (-1)*A`, sometimes does not hold. I'll come back to this.

Reply:
Hi. I'm extending Plutus Core to include some BLS operations based on this branch, and I'm pretty confident that `mod` is the right thing to use here. Scalar multiplication can be interpreted as repeated addition:
```
0p = 0
np = p + p + ... +p    (n times if n>0)
np = (-p) + (-p) + ... + (-p)   (-n times if n<0)
```
With these definitions you have sensible laws like n(p+q) = np + nq and (mn)p = m(np), and in fact earlier this week I added [tests](https://github.com/input-output-hk/plutus/blob/efd3d1ff5859abbd0abfb3e39bd702487b85532d/plutus-core/untyped-plutus-core/test/Evaluation/Builtins/BLS12_381/HaskellTests.hs#L81) to check such properties: they all pass, for all integral values (although there was an unrelated problem that we have to fix). I'm also paranoid about `rem` and `mod` (the fact that we have both is very confusing and I have to check the documentation every time I use one of those); I changed `mod` to `rem` in the code you quoted above and then the tests started failing for negative numbers ("point not on curve").  I think that this is because all of the functions that `blst` uses to construct scalars (like [this](https://github.com/supranational/blst/blob/d334254a4125cd2676d40b11229df208db8902e0/src/exports.c#L316) and the ones below) expect some unsigned representation of integers (although they don't actually say _what_ they expect anywhere), so if you supply a negative input the signed representation introduces unexpected bit patterns.

The crucial point here is that if we let N be the order of the group (ie, `scalarPeriod`), then `N*p = 0` for every point in G1 or G2, and it follows that (n+kN)*p = np for all integers k and n, positive or negative.  For example (-1)*p = (N-1)*p is an expensive way of calculating -p.  Now (as long as N>0),`a mod N` always gives you a result r with 0 <= r < N (which is thus a safe input to the `blst` scalar conversions), and `mod` also has the property that `(a +kN) mod n = a mod n`, so it's periodic modulo N in exactly the same way that repeated addition is (remember Np is always zero) and you get consistent results.

So I think it's perfectly safe to allow negative `Integer`s here. as long as everything's reduced modulo `scalarPeriod` before being fed to `blst` .  Maybe we don't _want_ to allow negative numbers, but I don't think it'd do any harm. You could argue that `blst` doesn't allow them, but then it also doesn't allow scalars bigger than `scalarPeriod`(although I think you can put any unsigned 256-bit value into a blst scalar, including some that are a bit bigger than the period: it'd be interesting to see what happens if you actually do that).

Reply:
> However, Now that I'm thinking about it, the `NEG` operation on an ecpoint might be cheaper (it's a sign bit flip only).

For G1, scalar multiplcation in Plutus is about 300 times more expensive than negation, and for G2 it's 600 times more expensive: you really don't want to be doing scalar multiplication if you can avoid it!

Reply:
>the fact that we have both is very confusing

In Haskell `rem` is a primitive operation, while `mod` is implemented in terms of `rem`. For that reason `rem` is more efficient and thus better when dealing with non-negative numbers only (eg. array indices)

> it's perfectly safe to allow negative Integers here. as long as everything's reduced modulo scalarPeriod before being fed to blst

That's exactly what current implementation of `scalarFromInteger` does, by using `mod x scalarPeriod` negative  numbers are converted to positive, so as far as `c_blst_p2_mult` and `c_blst_p1_mult` are concerned they will be getting positive numbers

> the NEG operation on an ecpoint might be cheaper (it's a sign bit flip only)

Negation doesn't look to be just a bit flip. If I had to guess it also looks like modulo arithmetic.

1. https://github.com/supranational/blst/blob/d334254a4125cd2676d40b11229df208db8902e0/src/e2.c#L91-L95
2. https://github.com/supranational/blst/blob/d334254a4125cd2676d40b11229df208db8902e0/src/fields.h#L78
3. https://github.com/supranational/blst/blob/d334254a4125cd2676d40b11229df208db8902e0/src/asm/add_mod_384-armv8.pl#L468-L516

> Maybe we don't want to allow negative numbers, but I don't think it'd do any harm.

Also my guess would be that this `mul (neg x) p` or this `neg (mul x p)` could be more efficient then `mul x (negate p)`. This is the only thing I can think of, if those are even correct properties that is :smile: 

In any case, this are just my thoughts. I am not even familiar with BLS12-381, so you should decide on how it should behave. Feel free to resolve this comment at any time

Reply:
> Negation doesn't look to be just a bit flip. If I had to guess it also looks like modulo arithmetic.

Ah, you are right. The bit flip works with compressed points. In compressed form only the sign of the `y` coordinate is stored, so a negation is simply changing that byte. However, with the full `y` coordinate, we indeed need the modular reduction. 

 > scalar multiplcation in Plutus is about 300 times more expensive than negation
 
 But none of the tricks here save us the scalar multiplication. I think we are really deciding of whether negating a number modulo ~32 bytes (negating a scalar) is faster than negating one modulo ~48 bytes (negating the `y` coordinate). I guess the difference is very small, particularly in comparison with the other operations. I'd leave it as it is, and let the user decide what it wants to do.

Reply:
I'll resolve this.